docs/source/main_classes/trainer.mdx

```diff
@@ -0,0 +1,550 @@
+
+For example here is how you could use it for `run_translation.py` with 2 GPUs:
+
+```bash
+python -m torch.distributed.launch --nproc_per_node=2 examples/pytorch/translation/run_translation.py \
+--model_name_or_path t5-small --per_device_train_batch_size 1   \
+--output_dir output_dir --overwrite_output_dir \
+--do_train --max_train_samples 500 --num_train_epochs 1 \
+--dataset_name wmt16 --dataset_config "ro-en" \
+--source_lang en --target_lang ro \
+--fp16 --sharded_ddp simple
+```
+
+Notes:
+
+- This feature requires distributed training (so multiple GPUs).
+- It is not implemented for TPUs.
+- It works with `--fp16` too, to make things even faster.
+- One of the main benefits of enabling `--sharded_ddp simple` is that it uses a lot less GPU memory, so you should be
+  able to use significantly larger batch sizes using the same hardware (e.g. 3x and even bigger) which should lead to
+  significantly shorter training time.
+
+3. To use the second version of Sharded data-parallelism, add `--sharded_ddp zero_dp_2` or `--sharded_ddp zero_dp_3` to the command line arguments, and make sure you have added the distributed launcher `-m torch.distributed.launch --nproc_per_node=NUMBER_OF_GPUS_YOU_HAVE` if you haven't been using it already.
+
+For example here is how you could use it for `run_translation.py` with 2 GPUs:
+
+```bash
+python -m torch.distributed.launch --nproc_per_node=2 examples/pytorch/translation/run_translation.py \
+--model_name_or_path t5-small --per_device_train_batch_size 1   \
+--output_dir output_dir --overwrite_output_dir \
+--do_train --max_train_samples 500 --num_train_epochs 1 \
+--dataset_name wmt16 --dataset_config "ro-en" \
+--source_lang en --target_lang ro \
+--fp16 --sharded_ddp zero_dp_2
+```
+
+`zero_dp_2` is an optimized version of the simple wrapper, while `zero_dp_3` fully shards model weights,
+gradients and optimizer states.
+
+Both are compatible with adding `cpu_offload` to enable ZeRO-offload (activate it like this: `--sharded_ddp "zero_dp_2 cpu_offload"`).
+
+Notes:
+
+- This feature requires distributed training (so multiple GPUs).
+- It is not implemented for TPUs.
+- It works with `--fp16` too, to make things even faster.
+- The `cpu_offload` additional option requires `--fp16`.
+- This is an area of active development, so make sure you have a source install of fairscale to use this feature as
+  some bugs you encounter may have been fixed there already.
+
+Known caveats:
+
+- This feature is incompatible with `--predict_with_generate` in the _run_translation.py_ script.
+- Using `--sharded_ddp zero_dp_3` requires wrapping each layer of the model in the special container
+  `FullyShardedDataParallelism` of fairscale. It should be used with the option `auto_wrap` if you are not
+  doing this yourself: `--sharded_ddp "zero_dp_3 auto_wrap"`.
+
+
+### DeepSpeed
+
+
+Moved to [Trainer DeepSpeed integration](deepspeed#trainer-deepspeed-integration).
+
```

Comment:
I kept everything as is @stas00 but maybe it's time to remove everything below and only leave this first line? The rendering on the doc is not super nice, as seen [here](https://huggingface.co/docs/transformers/main_classes/trainer#deepspeed)

Reply:
The intention was to preserve old links. so if someone on another site/forum linked to a specific section the user following this link will be able to find the new location after the move.

If you remove it, the anchor will fail and the user will get the top of the page and won't even know about that main: `moved to ...` 

One easy way to fix this on the web server level is to create a redirect file which matches the old link with the anchor and redirects it to a new location, .e.g;

```
https://huggingface.co/docs/transformers/main_classes/trainer#deepspeed| https://huggingface.co/docs/transformers/master/en/main_classes/trainer#deepspeed-trainer-integration
... same for all other redirects ...
```

we already have a redirect rule from `.html` to non-html link, so these need to be added there.

It appears hf.co is running on `nginx` webserver, so need to look up the format for redirects on that server.

then we can nuke those "poor man's redirects" in the doc itself.

Reply:
should we discuss this on #doc-frontend?

Reply:
Sure!

Reply:
nginx does not see anchor links, this is a purely client-side element

Reply:
Thank you, @julien-c, for sharing that. Indeed the browser doesn't send the anchor part to the server :(

That means we can't do the redirect in the rules file :(

It's your call Sylvain, if you prefer clean and neat over supporting existing links then please remove all but the first one.

Alternatively, we could discuss and agree on a general policy - when we rename headers or move content to a different file, do we support old links or not and at what cost. And depending on the outcome of that discussion then this decision about the current question can be made.

Reply:
> In this case, how about we just leave the HTML anchors and remove all visible content? A user that tries a "dead" link will then see "Moved to [Trainer DeepSpeed integration]" and can then easily find their way from there.

If it works that would be great. Could you do a single link mockup to see what you mean exactly?

Something like:
```
<h3><a href="#deepspeed"></a>DeepSpeed</h3>
<p>Moved to <a href="/docs/transformers/master/en/main_classes/trainer#deepspeed-trainer-integration">deepspeed-trainer-integration</a>.</p>
```

but add it to a non-visible class?

Reply:
I'll open a fresh PR to discuss this and merge this one in the meantime to fix the related issue :-)